### PR TITLE
Adding AstraDB as a DocumentStore

### PR DESCRIPTION
**Pull request by hc33brackles**
This is the PR discussed with Masci and Tuana which adds Astra using the JSON API as a DocumentStore for Haystack built for Haystack 2.0. 

All unit tests passing.

Working examples. 

Looking forward to feedback. :-) --Nick

### CODE REVIEW

File: integrations/astra/tests/test_document_store.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+import pytest
+from haystack import Document
+from haystack.document_stores import DuplicatePolicy, MissingDocumentError
+from haystack.testing.document_store import DocumentStoreBaseTests
+
+from astra_store.document_store import AstraDocumentStore
+
+
+class TestDocumentStore(DocumentStoreBaseTests):
```
Comment (julian-risch) [inline integrations/astra/tests/test_document_store.py:15]: We should also have at least minimal tests for the retriever and the filters. For example, you could have a look at pinecone: https://github.com/deepset-ai/haystack-core-integrations/blob/main/integrations/pinecone/tests/test_dense_retriever.py and https://github.com/deepset-ai/haystack-core-integrations/blob/main/integrations/pinecone/tests/test_filters.py

**Review by julian-risch — CHANGES_REQUESTED** [review]
Hello @hc33brackles Thank you for opening this PR. It looks quite good already! 👍 I left some comments after reviewing the code. However, I didn't have the time to run the examples yet. Please let me know if you have any questions regarding the review. 🙂

File: integrations/astra/tests/test_document_store.py (hunk 1 above)
Comment (hc33brackles) [reply integrations/astra/tests/test_document_store.py:15]: Can you clarify this point a little bit? @julian-risch

**Comment by julian-risch** [comment]
@hc33brackles By the way, in order to run the other CI steps, the Contributor License Agreement needs to be signed by all contributors. You can find the check above: https://github.com/deepset-ai/haystack-core-integrations/pull/144#issuecomment-1867779127

File: integrations/astra/src/astra_store/astra_client.py [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+class Response:
+    id: str
+    text: Optional[str]
+    values: Optional[list]
+    metadata: Optional[dict]
+    score: Optional[float]
+
+
+@dataclass
+class QueryResponse:
```
[the remark this answers is not in the document]
Comment (Briah) [reply integrations/astra/src/astra_store/astra_client.py:22]: Correct! This is an Astra internal data class

File: integrations/astra/tests/test_document_store.py (hunk 1 above)
Comment (anomnaco) [reply integrations/astra/tests/test_document_store.py:15]: We are running the same filter tests as Pinecone. Their test_filters.py inherits tests from FilterDocumentsTest in haystack/haystack/testing/document_store.py, while our test_document_store inherits its tests from DocumentStoreBaseTests which includes the tests from FilterDocumentsTest

File: integrations/astra/src/astra_store/document_store.py [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+
+    @validate_arguments
+    def __init__(
+        self,
+        astra_id: str,
+        astra_region: str,
+        astra_application_token: str,
+        astra_keyspace: str,
+        astra_collection: str,
+        embedding_dim: int,
```
[the remark this answers is not in the document]
Comment (hc33brackles) [reply integrations/astra/src/astra_store/document_store.py:45]: Our team had no problem with this :-)

File: integrations/astra/tests/test_document_store.py (hunk 1 above)
Comment (hc33brackles) [reply integrations/astra/tests/test_document_store.py:15]: We will be using the retriever test though

**Comment by hc33brackles** [comment]
@julian-risch - we are still responding to some comments but all contributors have signed!

**Comment by mounaTay** [comment]
@julian-risch We just pushed all the needed fixes and added our CI work as well.
Please let us know if you think we should add more changes to this PR
An FYI about the CI, it requires setting some git secrets to be able to connect to the datastax's database

**Comment by hc33brackles** [comment]
Mouna is working on the linting issues, but there might be something more underlying with the way we changed the astra secrets to be handled is causing some of the errors in the python components. @julian-risch - I think if you input the ASTRA_DB_APPLICATION_TOKEN and ASTRA_DB_ID to the test repo as repo secrets these tests will pass.

File: integrations/astra/tests/test_document_store.py [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+from typing import List
+
+import pytest
+from haystack import Document
+from haystack.document_stores import DuplicatePolicy, MissingDocumentError
+from haystack.testing.document_store import DocumentStoreBaseTests
+
+from astra_haystack.document_store import AstraDocumentStore
+
+
```
Comment (julian-risch) [inline integrations/astra/tests/test_document_store.py:13]:
```suggestion

@pytest.mark.skipif(os.environ.get("ASTRA_DB_APPLICATION_TOKEN", "") == "",
                    reason="ASTRA_DB_APPLICATION_TOKEN is not set")
@pytest.mark.skipif(os.environ.get("ASTRA_DB_ID", "") == "", reason="ASTRA_DB_ID is not set")
```

File: integrations/astra/tests/test_document_store.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
+# SPDX-FileCopyrightText: 2023-present Anant Corporation <support@anant.us>
+#
+# SPDX-License-Identifier: Apache-2.0
+from typing import List
```
Comment (julian-risch) [inline integrations/astra/tests/test_document_store.py:4]:
```suggestion
import os
from typing import List
```

File: integrations/astra/tests/test_retriever.py [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+#
+# SPDX-License-Identifier: Apache-2.0
+import os
+
+import pytest
+
+from astra_haystack.retriever import AstraRetriever
+
+
+@pytest.mark.integration
```
Comment (julian-risch) [inline integrations/astra/tests/test_retriever.py:11]:
```suggestion
@pytest.mark.skipif(os.environ.get("ASTRA_DB_APPLICATION_TOKEN", "") == "",
                    reason="ASTRA_DB_APPLICATION_TOKEN is not set")
@pytest.mark.skipif(os.environ.get("ASTRA_DB_ID", "") == "", reason="ASTRA_DB_ID is not set")
@pytest.mark.integration
```

File: integrations/astra/tests/test_retriever.py [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+                    "embedding_dim": 768,
+                    "similarity": "cosine",
+                },
+                "type": "astra_haystack.document_store.AstraDocumentStore",
+            },
+        },
+    }
+
+
+@pytest.mark.integration
```
Comment (julian-risch) [inline integrations/astra/tests/test_retriever.py:35]:
```suggestion
@pytest.mark.skipif(os.environ.get("ASTRA_DB_APPLICATION_TOKEN", "") == "",
                    reason="ASTRA_DB_APPLICATION_TOKEN is not set")
@pytest.mark.skipif(os.environ.get("ASTRA_DB_ID", "") == "", reason="ASTRA_DB_ID is not set")
@pytest.mark.integration
```

**Review by julian-risch — CHANGES_REQUESTED** [review]
Looks all very good to me! 👍 We just need to skip the tests when the API secret tokens are not available. I suggested the required changes so that you can commit them directly from the UI. I couldn't commit to your branch directly.
Thank you @mounaTay for your help running the tests locally and confirming the availability of the keys was an issue.
Once the changes I suggested with this review are committed the CI should pass and I will approve and merge the PR.

**Review by julian-risch — APPROVED** [review]
LGTM! Great job everyone involved!

**Pull request merged**